### ISSUE TITLE
Non-normalized AlignedAxis Is Ignored

### ISSUE THREAD

**Issue by esparano** [CesiumGS/cesium#6596]
The Billboard documentation states that "The aligned axis is the unit vector that the billboard up vector points towards. The default is the zero vector, which means the billboard is aligned to the screen up vector." This indicates that the aligned axis should be a unit vector. 

However, if you give the Billboard a non-normalized "unit" vector, that vector is completely ignored and the default is used instead. I think that either:
A) The documentation should clearly state that any non-normalized alignedAxis will be ignored and replaced with the Zero vector.
or better yet,
B) The alignedAxis should not have to be normalized in the first place. The Cesium code can normalize the axis internally if it needs to.

**Comment by shunter** [comment]
This sounded familiar so I searched through old issues to refresh my memory and summarize previous work on this topic.  

In #4046 I clarified the documentation to specify that aligned axis must always be a unit vector, while adding the ability to use the velocity vector as an aligned axis.
In #5807 the CZML loading code makes sure to always normalize unit cartesian values on load.  This was to avoid issues arising from imperfect floating point rounding errors when values are serialized to JSON (i.e. magnitude ends up being slightly above or below 1 at machine precision)

**Comment by esparano** [comment]
Yeah I was having a similar issue where the magnitude wasn't exactly 1, so it "silently failed" and made the billboards align with the camera instead of just normalizing the vector for me or giving a warning or something. I hope one of my two suggestions could help fix that type of silent failing.

**Comment by hpinkos** [comment]
Yeah, I think this should throw a `DeveloperError` instead of silently failing

```js
var viewer = new Cesium.Viewer('cesiumContainer');
viewer.entities.add({
    position : Cesium.Cartesian3.fromDegrees(-75.59777, 40.03883),
    billboard :{
        image : '../images/Cesium_Logo_overlay.png',
        alignedAxis: new Cesium.Cartesian3(0.0, 0.0, 2.0)
    }
});
```

**Comment by Chelsy-AI** [comment]
Hi, I'm a new contributor working through an open source contribution program. I'd like to work on this issue, specifically adding a DeveloperError when a non normalized alignedAxis is passed to a Billboard. Is this still available?

**Comment by javagl** [comment]
@Chelsy-AI Certainly. There have been two suggestions:

1. either throw a `DeveloperError` when it is not normalized
2. or normalize it automatically 

Disclaimer: I didn't have that `alignedAxis` on the radar, and don't know how commonly it is used. 

But considering the possible (side) effects:

The effect of 1. could be that existing applications (that pass in a non-normalized vector) would suddently stop working, because they'd throw that `DeveloperError`.

The effect of 2. could be that the billboards in existing applications could suddenly have "wrong" orientations (from the perspective of the application) **iff** the application _intentionally_ passed in non-unit-vectors, exploiting the "quirky" behavior that these are ignored. This is _very_ unlikely. (And the fact that they are ignored is not even specified anyhow...)

If I didn't miss anything here: 

I'd lean towards resolving this by automatically normalizing the axis (instead of throwing a `DeveloperError`), because it is less likely to have breaking/unexpected changes.

Now... we're between a rock and a hard place: 

The [documentation of `alignedAxis`](https://cesium.com/learn/ion-sdk/ref-doc/Billboard.html#alignedAxis) _explicitly_ says that it can be set to `Cartesian3.ZERO` to "reset the aligned axis". (Who came up with that one?). So this will require a special treatment. Usually, such cases are treated with ~"some epsilon-check for the length". But in this particular case, I'd consider to do an _explicit_ check for the length being `=== 0.0`.

A _quick(!)_ search indicates that [this](https://github.com/CesiumGS/cesium/blob/0e063e22b471d0fe08414d1d6ef81cfebd73177c/packages/engine/Source/Scene/BillboardCollection.js#L1099) _might_ be the place where this issue is caused: Any axis that is not-(nearly-) unit length seems to be deemed "invalid" and ignored. (This remains to be confirmed). 

In any case, giving the `Cartesian.ZERO` (i.e. the `length === 0` case) a special meaning seems to be appropriate in this case.

---

Just a sandcastle that shows the different behavior of the unit- and non-unit axes:

https://sandcastle.cesium.com/#c=vZPfS8MwEMf/ldCXdVLTDR3C7IbbfBz4IPhUGGl7q4dpUpJ0c8r+d5PF6n754IMWUi6X+3yv9yXFqpbKkAvCNJmBxqYiSyUrkgb5bpcGt6lKhVu5FNqQFcIaFBkRAetPgj7tcmHLzKQwDAWoNOge0QUqyA1KcSgwY8rYiImrsE97EWlfHvcoe0U9sdgJQoVUFeP4BqGrJt9Noh96dI91p2d1cy7FbzXd8hZREAYNgqasKMJ3L0NILTXuDBie6eisv4dSAejw8mbgXBj0rA1RS2fIeSaZKsjwS9E9WLESrGSH0ngX69irL+aylAu5AsXZhtai7ET7nLWtFFBMrAdD73B7unXB9o+Guv7foabHQwVRkGiz4TB2+Tv0P0GjeGhbGahqzuzXx1mTv4ChudbWB5HELZIUuCJYjM5ceJJzprU9WTacP9o7mQbjJLb1BxiXrEBRPvgJXMlzfzz3SUppEtvtKWWk5BlTe4of

**Comment by Chelsy-AI** [comment]
> [@Chelsy-AI](https://github.com/Chelsy-AI) Certainly. There have been two suggestions:
> 
> 1. either throw a `DeveloperError` when it is not normalized
> 2. or normalize it automatically
> 
> Disclaimer: I didn't have that `alignedAxis` on the radar, and don't know how commonly it is used.
> 
> But considering the possible (side) effects:
> 
> The effect of 1. could be that existing applications (that pass in a non-normalized vector) would suddently stop working, because they'd throw that `DeveloperError`.
> 
> The effect of 2. could be that the billboards in existing applications could suddenly have "wrong" orientations (from the perspective of the application) **iff** the application _intentionally_ passed in non-unit-vectors, exploiting the "quirky" behavior that these are ignored. This is _very_ unlikely. (And the fact that they are ignored is not even specified anyhow...)
> 
> If I didn't miss anything here:
> 
> I'd lean towards resolving this by automatically normalizing the axis (instead of throwing a `DeveloperError`), because it is less likely to have breaking/unexpected changes.
> 
> Now... we're between a rock and a hard place:
> 
> The [documentation of `alignedAxis`](https://cesium.com/learn/ion-sdk/ref-doc/Billboard.html#alignedAxis) _explicitly_ says that it can be set to `Cartesian3.ZERO` to "reset the aligned axis". (Who came up with that one?). So this will require a special treatment. Usually, such cases are treated with ~"some epsilon-check for the length". But in this particular case, I'd consider to do an _explicit_ check for the length being `=== 0.0`.
> 
> A _quick(!)_ search indicates that [this](https://github.com/CesiumGS/cesium/blob/0e063e22b471d0fe08414d1d6ef81cfebd73177c/packages/engine/Source/Scene/BillboardCollection.js#L1099) _might_ be the place where this issue is caused: Any axis that is not-(nearly-) unit length seems to be deemed "invalid" and ignored. (This remains to be confirmed).
> 
> In any case, giving the `Cartesian.ZERO` (i.e. the `length === 0` case) a special meaning seems to be appropriate in this case.
> 
> Just a sandcastle that shows the different behavior of the unit- and non-unit axes:
> 
> https://sandcastle.cesium.com/#c=vZPfS8MwEMf/ldCXdVLTDR3C7IbbfBz4IPhUGGl7q4dpUpJ0c8r+d5PF6n754IMWUi6X+3yv9yXFqpbKkAvCNJmBxqYiSyUrkgb5bpcGt6lKhVu5FNqQFcIaFBkRAetPgj7tcmHLzKQwDAWoNOge0QUqyA1KcSgwY8rYiImrsE97EWlfHvcoe0U9sdgJQoVUFeP4BqGrJt9Noh96dI91p2d1cy7FbzXd8hZREAYNgqasKMJ3L0NILTXuDBie6eisv4dSAejw8mbgXBj0rA1RS2fIeSaZKsjwS9E9WLESrGSH0ngX69irL+aylAu5AsXZhtai7ET7nLWtFFBMrAdD73B7unXB9o+Guv7foabHQwVRkGiz4TB2+Tv0P0GjeGhbGahqzuzXx1mTv4ChudbWB5HELZIUuCJYjM5ceJJzprU9WTacP9o7mQbjJLb1BxiXrEBRPvgJXMlzfzz3SUppEtvtKWWk5BlTe4of

Thank you for the detailed explanation! I'll go with option 2 — auto-normalizing the axis, with an explicit check for length === 0 to preserve the Cartesian3.ZERO reset behavior. I'll look at BillboardCollection.js around line 1099 as the starting point.

**Comment by javagl** [comment]
@RickyH22 Given that Chelsy-AI already commented and asked to work on this, maybe you want to have a look at a different issue (we only recently started 'assigning' people to issues, just to keep things a bit more organized during the FCF contributions)
